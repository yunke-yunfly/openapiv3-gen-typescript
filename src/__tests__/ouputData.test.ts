import {
  combineSingleApi,
  combineSingleType,
  getApiDTs,
  getApis,
  getCommonDTs,
  getOutputData,
} from '../outputData';

describe('outputData', () => {
  describe('getCommonDTs', () => {
    test('当无值是时候，返回 undefined', async () => {
      const res = await getCommonDTs(undefined);
      expect(res).toBe(undefined);
    });
    test('正常编译多个 definitions', async () => {
      const definitions: any = {
        Foo: {
          type: 'object',
          properties: {
            id: {
              type: 'string',
              description: '这是 ID',
            },
          },
          required: ['id'],
          additionalProperties: false,
        },
        Bar: {
          type: 'object',
          properties: {
            zoo: {
              $ref: '#/definitions/Foo',
            },
          },
          required: ['zoo'],
          additionalProperties: false,
        },
      };
      const res = await getCommonDTs(definitions);

      expect(res.fileName).toBe('common.d.ts');
      expect(res.content).toBe(`export interface Foo {
/**
 * 这是 ID
 */
id: string
}
export interface Bar {
zoo: Foo
}`);
    });
  });

  describe('combineSingleType', () => {
    test('有 response、fnComment、successResponseTemplate', () => {
      const api: any = {
        fnName: 'getUser',
        fnComment: `/**
                    * 我是注释
                    */\n`,
        body: `
          export interface Body {
            name: string;
          }
        `,
        query: `
          export interface Query {
            age: number;
          }
        `,
        response: `
          export interface Response {
            id: number
          }
        `,
      };
      expect(
        combineSingleType(api, { successResponseTemplate: `{code: number; data: T}` } as any),
      ).toMatchSnapshot();
    });

    test('无 response 且无 successResponseTemplate', () => {
      const api: any = {
        fnName: 'get_user',
        body: `
          export interface Body {
            name: string;
          }
        `,
      };
      expect(combineSingleType(api, {} as any)).toMatchSnapshot();
    });
  });

  describe('combineSingleApi', () => {
    test('参数齐全', () => {
      const api: any = {
        fnName: 'getUser',
        fnComment: `/**
                    * 我是注释
                    */\n`,
        body: `
          export interface Body {
            name: string;
          }
        `,
        query: `
          export interface Query {
            age: number;
          }
        `,
        header: `
          export interface Header {
            orgcode: number;
          }
        `,
        response: `
          export interface Response {
            id: number
          }
        `,
        params: `name: string, age: number`,
        method: 'POST',
        url: '/user/${name}',
      };

      expect(combineSingleApi(api)).toMatchSnapshot();
    });

    test('query、body、header、cookie、params 不存在时', () => {
      const api: any = {
        fnName: 'getUser',
        url: '/user',
      };

      expect(combineSingleApi(api)).toMatchSnapshot();
    });
  });

  describe('getApiDTs', () => {
    test('正常渲染', () => {
      const apis = [
        {
          fnName: 'getUser',
          fileName: 'user',
          url: '/user',
        },
        {
          fnName: 'createUser',
          fileName: 'user',
          url: '/user',
        },
        {
          fnName: 'createTeam',
          fileName: 'team',
          url: '/team',
        },
      ];

      expect(getApiDTs(apis as any, {} as any)).toMatchSnapshot();
    });

    test('包含 Types ', () => {
      const apis = [
        {
          fnName: 'getUser',
          fileName: 'user',
          url: '/user',
          response: `
            interface Data {
              userInfo: Types.UserInfo
            }
          `,
        },
      ];

      expect(getApiDTs(apis as any, {} as any)).toMatchSnapshot();
    });
  });

  describe('getApis', () => {
    test('正常渲染', () => {
      const apis = [
        {
          fnName: 'getUser',
          fileName: 'user',
          url: '/user',
        },
        {
          fnName: 'createUser',
          fileName: 'user',
          url: '/user',
        },
        {
          fnName: 'createTeam',
          fileName: 'team',
          url: '/team',
        },
      ];

      expect(
        getApis(
          apis as any,
          { requestTemplate: 'import { request } from "../utils/request"' } as any,
        ),
      ).toMatchSnapshot();
    });

    test('config.requestTemplate 无值', () => {
      const apis = [
        {
          fnName: 'getUser',
          fileName: 'user',
          url: '/user',
        },
      ];

      expect(getApis(apis as any, { axiosFilePath: '../utils/request' } as any)).toMatchSnapshot();
    });
  });

  describe('getOutputData', () => {
    test('有 definitions', async () => {
      const apis: any = [
        {
          fnName: 'getUser',
          fileName: 'user',
          url: '/user',
        },
      ];

      const definitions = {
        Foo: {
          type: 'object',
          properties: {
            id: {
              type: 'string',
              description: '这是 ID',
            },
          },
          required: ['id'],
          additionalProperties: false,
        },
      };
      const res = await getOutputData(apis, definitions as any, {} as any);
      expect(res).toMatchSnapshot();
    });

    test('无 definitions', async () => {
      const apis: any = [
        {
          fnName: 'getUser',
          fileName: 'user',
          url: '/user',
        },
      ];
      const res = await getOutputData(apis, {} as any, {} as any);
      expect(res).toMatchSnapshot();
    });
  });
});
