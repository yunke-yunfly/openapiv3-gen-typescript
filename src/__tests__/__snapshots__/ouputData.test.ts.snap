// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`outputData combineSingleApi query、body、header、cookie、params 不存在时 1`] = `
"export function getUser(): Api.GetUser.Response {
  return request({
    method: 'GET',
    url: '/user',
    
  })
}
"
`;

exports[`outputData combineSingleApi 参数齐全 1`] = `
"/**
                    * 我是注释
                    */
export function getUser(name: string, age: number, query: Api.GetUser.Query, body: Api.GetUser.Body, header: Api.GetUser.Header): Api.GetUser.Response {
  return request({
    method: 'POST',
    url: \`/user/\${name}\`,
    params: query,
    data: body,
    headers: header
  })
}
"
`;

exports[`outputData combineSingleType 无 response 且无 successResponseTemplate 1`] = `
"export namespace GetUser {
      
          export interface Body {
            name: string;
          }
        
      export type Data = any
      export type Response = Promise<Data>
    }
  "
`;

exports[`outputData combineSingleType 有 response、fnComment、successResponseTemplate 1`] = `
"/**
                    * 我是注释
                    */
export namespace GetUser {
      
          export interface Query {
            age: number;
          }
        
  
          export interface Body {
            name: string;
          }
        
      
          export interface Response {
            id: number
          }
        
      export type Response = Promise<ApiResponse<Data>>
    }
  "
`;

exports[`outputData getApiDTs 包含 Types  1`] = `
Array [
  Object {
    "content": "export * from './types/user'",
    "fileName": "./types.d.ts",
  },
  Object {
    "content": "import Types from '../common';
export namespace User {
 export namespace GetUser {
      
      
            interface Data {
              userInfo: Types.UserInfo
            }
          
      export type Response = Promise<Data>
    }
  
}
export default User",
    "fileName": "./types/user.d.ts",
  },
]
`;

exports[`outputData getApiDTs 正常渲染 1`] = `
Array [
  Object {
    "content": "export * from './types/user'
export * from './types/team'",
    "fileName": "./types.d.ts",
  },
  Object {
    "content": "export namespace User {
 export namespace GetUser {
      
      export type Data = any
      export type Response = Promise<Data>
    }
  export namespace CreateUser {
      
      export type Data = any
      export type Response = Promise<Data>
    }
  
}
export default User",
    "fileName": "./types/user.d.ts",
  },
  Object {
    "content": "export namespace Team {
 export namespace CreateTeam {
      
      export type Data = any
      export type Response = Promise<Data>
    }
  
}
export default Team",
    "fileName": "./types/team.d.ts",
  },
]
`;

exports[`outputData getApis config.requestTemplate 无值 1`] = `
Array [
  Object {
    "content": "import request from '../utils/request'
import Api from '../types/user'
export function getUser(): Api.GetUser.Response {
  return request({
    method: 'GET',
    url: '/user',
    
  })
}

",
    "fileName": "./apis/user.ts",
  },
]
`;

exports[`outputData getApis 正常渲染 1`] = `
Array [
  Object {
    "content": "import { request } from \\"../utils/request\\"
import Api from '../types/user'
export function getUser(): Api.GetUser.Response {
  return request({
    method: 'GET',
    url: '/user',
    
  })
}

export function createUser(): Api.CreateUser.Response {
  return request({
    method: 'GET',
    url: '/user',
    
  })
}

",
    "fileName": "./apis/user.ts",
  },
  Object {
    "content": "import { request } from \\"../utils/request\\"
import Api from '../types/team'
export function createTeam(): Api.CreateTeam.Response {
  return request({
    method: 'GET',
    url: '/team',
    
  })
}

",
    "fileName": "./apis/team.ts",
  },
]
`;

exports[`outputData getOutputData 无 definitions 1`] = `
Array [
  Object {
    "content": "export * from './types/user'",
    "fileName": "./types.d.ts",
  },
  Object {
    "content": "export namespace User {
 export namespace GetUser {
      
      export type Data = any
      export type Response = Promise<Data>
    }
  
}
export default User",
    "fileName": "./types/user.d.ts",
  },
  Object {
    "content": "import request from 'undefined'
import Api from '../types/user'
export function getUser(): Api.GetUser.Response {
  return request({
    method: 'GET',
    url: '/user',
    
  })
}

",
    "fileName": "./apis/user.ts",
  },
]
`;

exports[`outputData getOutputData 有 definitions 1`] = `
Array [
  Object {
    "content": "export interface Foo {
/**
 * 这是 ID
 */
id: string
}",
    "fileName": "common.d.ts",
  },
  Object {
    "content": "export * from './types/user'",
    "fileName": "./types.d.ts",
  },
  Object {
    "content": "export namespace User {
 export namespace GetUser {
      
      export type Data = any
      export type Response = Promise<Data>
    }
  
}
export default User",
    "fileName": "./types/user.d.ts",
  },
  Object {
    "content": "import request from 'undefined'
import Api from '../types/user'
export function getUser(): Api.GetUser.Response {
  return request({
    method: 'GET',
    url: '/user',
    
  })
}

",
    "fileName": "./apis/user.ts",
  },
]
`;
